File: histogram_gui.py
"""
==========================================================
Program Histogram Specification (Peningkatan Mutu Citra)
Pengolahan Citra Digital - Python (Tkinter GUI)
==========================================================
Mendukung citra BERWARNA (RGB) - proses per channel R, G, B
==========================================================
"""

import tkinter as tk
from tkinter import ttk, filedialog, messagebox
from PIL import Image, ImageTk
import numpy as np
import matplotlib
matplotlib.use('TkAgg')
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
from matplotlib.figure import Figure
import os


class HistogramApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Histogram Specification - Pengolahan Citra Digital")
        self.root.geometry("1400x880")
        self.root.configure(bg="#1e1e2e")
        self.root.minsize(1200, 750)

        # Data
        self.img_original = None       # numpy array RGB
        self.img_reference = None      # numpy array RGB
        self.img_equalized = None
        self.img_specified = None

        self._build_styles()
        self._build_ui()

    def _build_styles(self):
        style = ttk.Style()
        style.theme_use('clam')
        style.configure("Title.TLabel", font=("Segoe UI", 18, "bold"),
                         foreground="#cdd6f4", background="#1e1e2e")
        style.configure("Sub.TLabel", font=("Segoe UI", 10),
                         foreground="#a6adc8", background="#1e1e2e")
        style.configure("Main.TFrame", background="#1e1e2e")
        style.configure("Accent.TButton", font=("Segoe UI", 10, "bold"),
                         foreground="#1e1e2e", background="#89b4fa",
                         padding=(14, 8))
        style.map("Accent.TButton", background=[('active', '#74c7ec')])
        style.configure("Status.TLabel", font=("Segoe UI", 9),
                         foreground="#a6adc8", background="#181825",
                         padding=(8, 4))

    def _build_ui(self):
        # Header
        header = ttk.Frame(self.root, style="Main.TFrame")
        header.pack(fill=tk.X, padx=20, pady=(15, 5))
        ttk.Label(header, text="📊 Histogram Specification (RGB)",
                  style="Title.TLabel").pack(side=tk.LEFT)
        ttk.Label(header, text="Peningkatan Mutu Citra Digital - Berwarna",
                  style="Sub.TLabel").pack(side=tk.LEFT, padx=(15, 0))

        # Toolbar
        toolbar = ttk.Frame(self.root, style="Main.TFrame")
        toolbar.pack(fill=tk.X, padx=20, pady=8)

        ttk.Button(toolbar, text="📂 Load Citra Asli",
                   style="Accent.TButton",
                   command=self.load_original).pack(side=tk.LEFT, padx=(0, 8))
        ttk.Button(toolbar, text="📂 Load Citra Referensi",
                   style="Accent.TButton",
                   command=self.load_reference).pack(side=tk.LEFT, padx=(0, 8))
        ttk.Button(toolbar, text="⚡ Histogram Equalization",
                   style="Accent.TButton",
                   command=self.do_equalization).pack(side=tk.LEFT, padx=(0, 8))
        ttk.Button(toolbar, text="🎯 Histogram Specification",
                   style="Accent.TButton",
                   command=self.do_specification).pack(side=tk.LEFT, padx=(0, 8))
        ttk.Button(toolbar, text="💾 Simpan Hasil",
                   style="Accent.TButton",
                   command=self.save_result).pack(side=tk.LEFT, padx=(0, 8))
        ttk.Button(toolbar, text="🔄 Reset",
                   style="Accent.TButton",
                   command=self.reset_all).pack(side=tk.RIGHT)

        # Main content
        self.main_frame = ttk.Frame(self.root, style="Main.TFrame")
        self.main_frame.pack(fill=tk.BOTH, expand=True, padx=20, pady=5)

        self.fig = Figure(figsize=(14, 7), facecolor='#1e1e2e')
        self.fig.subplots_adjust(hspace=0.5, wspace=0.35)
        self.canvas = FigureCanvasTkAgg(self.fig, master=self.main_frame)
        self.canvas.get_tk_widget().pack(fill=tk.BOTH, expand=True)

        # Status bar
        self.status_var = tk.StringVar(value="Siap. Silakan load citra asli untuk memulai.")
        ttk.Label(self.root, textvariable=self.status_var,
                  style="Status.TLabel", anchor=tk.W).pack(fill=tk.X, side=tk.BOTTOM)

        self._show_welcome()

    def _show_welcome(self):
        self.fig.clear()
        ax = self.fig.add_subplot(111)
        ax.set_facecolor('#1e1e2e')
        ax.text(0.5, 0.55, "Histogram Specification (RGB)", fontsize=28,
                ha='center', va='center', color='#89b4fa', fontweight='bold',
                transform=ax.transAxes)
        ax.text(0.5, 0.40, "Mendukung Citra Berwarna (RGB)", fontsize=14,
                ha='center', va='center', color='#a6adc8', transform=ax.transAxes)
        ax.text(0.5, 0.28, "Klik 'Load Citra Asli' untuk memulai", fontsize=11,
                ha='center', va='center', color='#585b70', transform=ax.transAxes)
        ax.axis('off')
        self.canvas.draw()

    # ================ Helpers ================

    def _compute_cdf(self, channel):
        """Hitung CDF ternormalisasi dari satu channel."""
        hist, _ = np.histogram(channel.flatten(), bins=256, range=(0, 256))
        cdf = hist.cumsum()
        cdf_min = cdf[cdf > 0].min()
        cdf_normalized = (cdf - cdf_min) * 255 / (cdf.max() - cdf_min)
        return np.clip(cdf_normalized, 0, 255).astype(np.uint8), hist

    def _equalize_channel(self, channel):
        cdf, _ = self._compute_cdf(channel)
        return cdf[channel]

    def _specify_channel(self, src_channel, ref_channel):
        cdf_src, _ = self._compute_cdf(src_channel)
        cdf_ref, _ = self._compute_cdf(ref_channel)
        mapping = np.zeros(256, dtype=np.uint8)
        for i in range(256):
            diff = np.abs(cdf_ref.astype(int) - int(cdf_src[i]))
            mapping[i] = np.argmin(diff)
        return mapping[src_channel]

    def _ensure_rgb(self, img_array):
        """Pastikan array adalah RGB (3 channel)."""
        if len(img_array.shape) == 2:
            return np.stack([img_array]*3, axis=-1)
        if img_array.shape[2] == 4:  # RGBA
            return img_array[:, :, :3]
        return img_array

    def _plot_rgb_histogram(self, ax, img, title_text):
        """Plot histogram RGB overlay di satu axes."""
        colors = ['#ff6b6b', '#51cf66', '#339af0']  # R, G, B
        labels = ['Red', 'Green', 'Blue']
        for i, (c, lbl) in enumerate(zip(colors, labels)):
            ax.hist(img[:, :, i].flatten(), bins=256, range=(0, 256),
                    color=c, alpha=0.5, label=lbl, edgecolor='none')
        ax.set_title(title_text, color='#cdd6f4', fontsize=10, fontweight='bold')
        ax.legend(fontsize=7, loc='upper right',
                  facecolor='#313244', edgecolor='#45475a', labelcolor='#cdd6f4')
        ax.set_facecolor('#313244')
        ax.tick_params(colors='#a6adc8', labelsize=7)
        ax.set_xlabel("Intensitas", color='#a6adc8', fontsize=8)
        ax.set_ylabel("Frekuensi", color='#a6adc8', fontsize=8)
        for s in ax.spines.values():
            s.set_color('#45475a')

    # ================ File Loading ================

    def load_original(self):
        path = filedialog.askopenfilename(
            title="Pilih Citra Asli",
            filetypes=[("Image Files", "*.png *.jpg *.jpeg *.bmp *.tif *.tiff")])
        if not path:
            return
        try:
            img = np.array(Image.open(path))
            self.img_original = self._ensure_rgb(img)
            self.img_equalized = None
            self.img_specified = None
            self.status_var.set(f"Citra asli dimuat: {os.path.basename(path)} "
                                f"({self.img_original.shape[1]}x{self.img_original.shape[0]}, RGB)")
            self._display_original()
        except Exception as e:
            messagebox.showerror("Error", f"Gagal membuka citra:\n{e}")

    def load_reference(self):
        path = filedialog.askopenfilename(
            title="Pilih Citra Referensi",
            filetypes=[("Image Files", "*.png *.jpg *.jpeg *.bmp *.tif *.tiff")])
        if not path:
            return
        try:
            img = np.array(Image.open(path))
            self.img_reference = self._ensure_rgb(img)
            self.status_var.set(f"Citra referensi dimuat: {os.path.basename(path)}")
            self._display_with_reference()
        except Exception as e:
            messagebox.showerror("Error", f"Gagal membuka citra referensi:\n{e}")

    # ================ Processing ================

    def do_equalization(self):
        if self.img_original is None:
            messagebox.showwarning("Peringatan", "Silakan load citra asli terlebih dahulu!")
            return
        result = np.zeros_like(self.img_original)
        for ch in range(3):
            result[:, :, ch] = self._equalize_channel(self.img_original[:, :, ch])
        self.img_equalized = result
        self.status_var.set("Histogram Equalization selesai (RGB).")
        self._display_equalization_result()

    def do_specification(self):
        if self.img_original is None:
            messagebox.showwarning("Peringatan", "Silakan load citra asli terlebih dahulu!")
            return
        if self.img_reference is None:
            messagebox.showwarning("Peringatan", "Silakan load citra referensi terlebih dahulu!")
            return
        result = np.zeros_like(self.img_original)
        for ch in range(3):
            result[:, :, ch] = self._specify_channel(
                self.img_original[:, :, ch], self.img_reference[:, :, ch])
        self.img_specified = result
        self.status_var.set("Histogram Specification selesai (RGB).")
        self._display_specification_result()

    def save_result(self):
        img = self.img_specified if self.img_specified is not None else self.img_equalized
        if img is None:
            messagebox.showwarning("Peringatan", "Belum ada hasil untuk disimpan!")
            return
        path = filedialog.asksaveasfilename(
            title="Simpan Hasil", defaultextension=".png",
            filetypes=[("PNG", "*.png"), ("JPEG", "*.jpg"), ("BMP", "*.bmp")])
        if path:
            Image.fromarray(img).save(path)
            self.status_var.set(f"Hasil disimpan: {path}")

    # ================ Display ================

    def _display_original(self):
        self.fig.clear()
        ax1 = self.fig.add_subplot(1, 2, 1)
        ax1.imshow(self.img_original)
        ax1.set_title("Citra Asli (RGB)", color='#cdd6f4', fontsize=12, fontweight='bold')
        ax1.axis('off')

        ax2 = self.fig.add_subplot(1, 2, 2)
        self._plot_rgb_histogram(ax2, self.img_original, "Histogram Citra Asli (RGB)")
        self.canvas.draw()

    def _display_with_reference(self):
        if self.img_original is None:
            return
        self.fig.clear()

        ax1 = self.fig.add_subplot(2, 2, 1)
        ax1.imshow(self.img_original); ax1.set_title("Citra Asli", color='#cdd6f4', fontsize=10, fontweight='bold'); ax1.axis('off')

        ax2 = self.fig.add_subplot(2, 2, 2)
        self._plot_rgb_histogram(ax2, self.img_original, "Histogram Asli")

        ax3 = self.fig.add_subplot(2, 2, 3)
        ax3.imshow(self.img_reference); ax3.set_title("Citra Referensi", color='#cdd6f4', fontsize=10, fontweight='bold'); ax3.axis('off')

        ax4 = self.fig.add_subplot(2, 2, 4)
        self._plot_rgb_histogram(ax4, self.img_reference, "Histogram Referensi")
        self.canvas.draw()

    def _display_equalization_result(self):
        self.fig.clear()

        ax1 = self.fig.add_subplot(2, 2, 1)
        ax1.imshow(self.img_original); ax1.set_title("Citra Asli", color='#cdd6f4', fontsize=10, fontweight='bold'); ax1.axis('off')

        ax2 = self.fig.add_subplot(2, 2, 2)
        self._plot_rgb_histogram(ax2, self.img_original, "Histogram Asli")

        ax3 = self.fig.add_subplot(2, 2, 3)
        ax3.imshow(self.img_equalized); ax3.set_title("Hasil Equalization", color='#cdd6f4', fontsize=10, fontweight='bold'); ax3.axis('off')

        ax4 = self.fig.add_subplot(2, 2, 4)
        self._plot_rgb_histogram(ax4, self.img_equalized, "Histogram Equalized")
        self.canvas.draw()

    def _display_specification_result(self):
        self.fig.clear()

        ax1 = self.fig.add_subplot(2, 3, 1)
        ax1.imshow(self.img_original); ax1.set_title("Citra Asli", color='#cdd6f4', fontsize=9, fontweight='bold'); ax1.axis('off')

        ax2 = self.fig.add_subplot(2, 3, 2)
        ax2.imshow(self.img_reference); ax2.set_title("Citra Referensi", color='#cdd6f4', fontsize=9, fontweight='bold'); ax2.axis('off')

        ax3 = self.fig.add_subplot(2, 3, 3)
        ax3.imshow(self.img_specified); ax3.set_title("Hasil Specification", color='#cdd6f4', fontsize=9, fontweight='bold'); ax3.axis('off')

        ax4 = self.fig.add_subplot(2, 3, 4)
        self._plot_rgb_histogram(ax4, self.img_original, "Histogram Asli")

        ax5 = self.fig.add_subplot(2, 3, 5)
        self._plot_rgb_histogram(ax5, self.img_reference, "Histogram Referensi")

        ax6 = self.fig.add_subplot(2, 3, 6)
        self._plot_rgb_histogram(ax6, self.img_specified, "Histogram Hasil")
        self.canvas.draw()

    def reset_all(self):
        self.img_original = None
        self.img_reference = None
        self.img_equalized = None
        self.img_specified = None
        self.status_var.set("Siap. Silakan load citra asli untuk memulai.")
        self._show_welcome()


def main():
    root = tk.Tk()
    app = HistogramApp(root)
    root.mainloop()


if __name__ == "__main__":
    main()

File: histogram_gui.m
function histogram_gui()
% ==========================================================
% Program Histogram Specification (Peningkatan Mutu Citra)
% Pengolahan Citra Digital - MATLAB GUI
% ==========================================================
% Mendukung citra BERWARNA (RGB) - proses per channel R, G, B
% ==========================================================

    % ---- Buat Figure Utama ----
    fig = figure('Name', 'Histogram Specification (RGB) - PCD', ...
                 'NumberTitle', 'off', ...
                 'Position', [80 80 1300 750], ...
                 'MenuBar', 'none', ...
                 'ToolBar', 'none', ...
                 'Color', [0.12 0.12 0.18], ...
                 'Resize', 'on');
    
    % ---- Data Storage ----
    data.img_original = [];   % RGB
    data.img_reference = [];  % RGB
    data.img_equalized = [];
    data.img_specified = [];
    guidata(fig, data);
    
    % ---- UI Colors ----
    btn_color = [0.54 0.71 0.98];
    btn_fg = [0.12 0.12 0.18];
    lbl_fg = [0.80 0.84 0.96];
    lbl_bg = [0.12 0.12 0.18];
    
    % ---- Title ----
    uicontrol('Parent', fig, 'Style', 'text', ...
              'String', 'Histogram Specification (RGB) - Pengolahan Citra Digital', ...
              'Position', [20 700 550 35], ...
              'FontSize', 14, 'FontWeight', 'bold', ...
              'ForegroundColor', lbl_fg, 'BackgroundColor', lbl_bg, ...
              'HorizontalAlignment', 'left');
    
    % ---- Buttons ----
    uicontrol('Style', 'pushbutton', 'String', 'Load Citra Asli', ...
              'Position', [20 660 160 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', btn_color, ...
              'Callback', @btn_load_original);
    
    uicontrol('Style', 'pushbutton', 'String', 'Load Citra Referensi', ...
              'Position', [190 660 180 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', btn_color, ...
              'Callback', @btn_load_reference);
    
    uicontrol('Style', 'pushbutton', 'String', 'Histogram Equalization', ...
              'Position', [380 660 190 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', [0.65 0.89 0.63], ...
              'Callback', @btn_equalization);
    
    uicontrol('Style', 'pushbutton', 'String', 'Histogram Specification', ...
              'Position', [580 660 190 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', [0.95 0.55 0.66], ...
              'Callback', @btn_specification);
    
    uicontrol('Style', 'pushbutton', 'String', 'Simpan Hasil', ...
              'Position', [780 660 130 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', btn_color, ...
              'Callback', @btn_save);
    
    uicontrol('Style', 'pushbutton', 'String', 'Reset', ...
              'Position', [1180 660 100 35], 'FontSize', 10, 'FontWeight', 'bold', ...
              'ForegroundColor', btn_fg, 'BackgroundColor', [0.98 0.70 0.53], ...
              'Callback', @btn_reset);
    
    % Status
    status_label = uicontrol('Style', 'text', ...
              'String', 'Siap. Silakan load citra asli untuk memulai.', ...
              'Position', [20 10 800 22], 'FontSize', 9, ...
              'ForegroundColor', [0.65 0.68 0.78], ...
              'BackgroundColor', [0.09 0.09 0.15], ...
              'HorizontalAlignment', 'left');
    
    % ===== HELPER: Ensure RGB =====
    function img = ensure_rgb(img)
        if size(img, 3) == 1
            img = cat(3, img, img, img);
        end
    end
    
    % ===== HELPER: Plot RGB Histogram =====
    function plot_rgb_hist(ax, img, ttl)
        hold(ax, 'on');
        [countsR, ~] = imhist(img(:,:,1));
        [countsG, ~] = imhist(img(:,:,2));
        [countsB, binsB] = imhist(img(:,:,3));
        bar(ax, binsB, countsR, 'FaceColor', [1 0.4 0.4], 'FaceAlpha', 0.5, 'EdgeColor', 'none');
        bar(ax, binsB, countsG, 'FaceColor', [0.3 0.8 0.4], 'FaceAlpha', 0.5, 'EdgeColor', 'none');
        bar(ax, binsB, countsB, 'FaceColor', [0.2 0.6 0.9], 'FaceAlpha', 0.5, 'EdgeColor', 'none');
        hold(ax, 'off');
        title(ax, ttl, 'Color', lbl_fg, 'FontSize', 10);
        legend(ax, 'R', 'G', 'B', 'TextColor', lbl_fg, 'Color', [0.19 0.19 0.27], ...
               'EdgeColor', [0.27 0.28 0.35], 'FontSize', 7);
        ax.Color = [0.19 0.19 0.27];
        ax.XColor = [0.65 0.68 0.78];
        ax.YColor = [0.65 0.68 0.78];
        xlim(ax, [0 255]);
    end
    
    % ===== HELPER: Equalize one channel =====
    function out = eq_channel(ch)
        out = histeq(ch);
    end
    
    % ===== HELPER: Specify one channel =====
    function out = spec_channel(src_ch, ref_ch)
        out = imhistmatch(src_ch, ref_ch);
    end
    
    % ===== CALLBACKS =====
    
    function btn_load_original(~, ~)
        [file, path] = uigetfile({'*.png;*.jpg;*.jpeg;*.bmp;*.tif;*.tiff', ...
                                   'Image Files'}, 'Pilih Citra Asli');
        if isequal(file, 0), return; end
        
        d = guidata(fig);
        img = imread(fullfile(path, file));
        d.img_original = ensure_rgb(img);
        d.img_equalized = [];
        d.img_specified = [];
        guidata(fig, d);
        
        set(status_label, 'String', ['Citra asli dimuat (RGB): ' file]);
        show_original(d);
    end
    
    function btn_load_reference(~, ~)
        [file, path] = uigetfile({'*.png;*.jpg;*.jpeg;*.bmp;*.tif;*.tiff', ...
                                   'Image Files'}, 'Pilih Citra Referensi');
        if isequal(file, 0), return; end
        
        d = guidata(fig);
        img = imread(fullfile(path, file));
        d.img_reference = ensure_rgb(img);
        guidata(fig, d);
        
        set(status_label, 'String', ['Citra referensi dimuat (RGB): ' file]);
        show_with_reference(d);
    end
    
    function btn_equalization(~, ~)
        d = guidata(fig);
        if isempty(d.img_original)
            msgbox('Silakan load citra asli terlebih dahulu!', 'Peringatan', 'warn');
            return;
        end
        
        % Equalize per channel
        result = uint8(zeros(size(d.img_original)));
        for ch = 1:3
            result(:,:,ch) = eq_channel(d.img_original(:,:,ch));
        end
        d.img_equalized = result;
        guidata(fig, d);
        
        set(status_label, 'String', 'Histogram Equalization selesai (RGB).');
        show_equalization(d);
    end
    
    function btn_specification(~, ~)
        d = guidata(fig);
        if isempty(d.img_original)
            msgbox('Silakan load citra asli terlebih dahulu!', 'Peringatan', 'warn');
            return;
        end
        if isempty(d.img_reference)
            msgbox('Silakan load citra referensi terlebih dahulu!', 'Peringatan', 'warn');
            return;
        end
        
        % Specification per channel
        result = uint8(zeros(size(d.img_original)));
        for ch = 1:3
            result(:,:,ch) = spec_channel(d.img_original(:,:,ch), d.img_reference(:,:,ch));
        end
        d.img_specified = result;
        guidata(fig, d);
        
        set(status_label, 'String', 'Histogram Specification selesai (RGB).');
        show_specification(d);
    end
    
    function btn_save(~, ~)
        d = guidata(fig);
        if ~isempty(d.img_specified)
            img_out = d.img_specified;
        elseif ~isempty(d.img_equalized)
            img_out = d.img_equalized;
        else
            msgbox('Belum ada hasil untuk disimpan!', 'Peringatan', 'warn');
            return;
        end
        [file, path] = uiputfile({'*.png','PNG';'*.jpg','JPEG';'*.bmp','BMP'}, 'Simpan Hasil');
        if ~isequal(file, 0)
            imwrite(img_out, fullfile(path, file));
            set(status_label, 'String', ['Hasil disimpan: ' fullfile(path, file)]);
        end
    end
    
    function btn_reset(~, ~)
        d.img_original = [];
        d.img_reference = [];
        d.img_equalized = [];
        d.img_specified = [];
        guidata(fig, d);
        allAxes = findall(fig, 'Type', 'axes');
        delete(allAxes);
        set(status_label, 'String', 'Siap. Silakan load citra asli untuk memulai.');
    end
    
    % ===== DISPLAY =====
    
    function show_original(d)
        allAxes = findall(fig, 'Type', 'axes');
        delete(allAxes);
        
        ax1 = axes('Parent', fig, 'Position', [0.05 0.12 0.42 0.65]);
        imshow(d.img_original, 'Parent', ax1);
        title(ax1, 'Citra Asli (RGB)', 'Color', lbl_fg, 'FontSize', 12, 'FontWeight', 'bold');
        
        ax2 = axes('Parent', fig, 'Position', [0.55 0.12 0.42 0.65]);
        plot_rgb_hist(ax2, d.img_original, 'Histogram Citra Asli (RGB)');
    end
    
    function show_with_reference(d)
        if isempty(d.img_original), return; end
        allAxes = findall(fig, 'Type', 'axes');
        delete(allAxes);
        
        ax1 = axes('Parent', fig, 'Position', [0.03 0.45 0.22 0.40]);
        imshow(d.img_original, 'Parent', ax1);
        title(ax1, 'Citra Asli', 'Color', lbl_fg, 'FontSize', 10);
        
        ax2 = axes('Parent', fig, 'Position', [0.28 0.45 0.22 0.40]);
        plot_rgb_hist(ax2, d.img_original, 'Histogram Asli');
        
        ax3 = axes('Parent', fig, 'Position', [0.53 0.45 0.22 0.40]);
        imshow(d.img_reference, 'Parent', ax3);
        title(ax3, 'Citra Referensi', 'Color', lbl_fg, 'FontSize', 10);
        
        ax4 = axes('Parent', fig, 'Position', [0.78 0.45 0.20 0.40]);
        plot_rgb_hist(ax4, d.img_reference, 'Histogram Referensi');
    end
    
    function show_equalization(d)
        allAxes = findall(fig, 'Type', 'axes');
        delete(allAxes);
        
        ax1 = axes('Parent', fig, 'Position', [0.03 0.45 0.22 0.40]);
        imshow(d.img_original, 'Parent', ax1);
        title(ax1, 'Citra Asli', 'Color', lbl_fg, 'FontSize', 10);
        
        ax2 = axes('Parent', fig, 'Position', [0.28 0.45 0.22 0.40]);
        plot_rgb_hist(ax2, d.img_original, 'Histogram Asli');
        
        ax3 = axes('Parent', fig, 'Position', [0.53 0.45 0.22 0.40]);
        imshow(d.img_equalized, 'Parent', ax3);
        title(ax3, 'Hasil Equalization', 'Color', lbl_fg, 'FontSize', 10);
        
        ax4 = axes('Parent', fig, 'Position', [0.78 0.45 0.20 0.40]);
        plot_rgb_hist(ax4, d.img_equalized, 'Histogram Equalized');
    end
    
    function show_specification(d)
        allAxes = findall(fig, 'Type', 'axes');
        delete(allAxes);
        
        ax1 = axes('Parent', fig, 'Position', [0.03 0.50 0.30 0.38]);
        imshow(d.img_original, 'Parent', ax1);
        title(ax1, 'Citra Asli', 'Color', lbl_fg, 'FontSize', 10);
        
        ax2 = axes('Parent', fig, 'Position', [0.36 0.50 0.30 0.38]);
        imshow(d.img_reference, 'Parent', ax2);
        title(ax2, 'Citra Referensi', 'Color', lbl_fg, 'FontSize', 10);
        
        ax3 = axes('Parent', fig, 'Position', [0.69 0.50 0.30 0.38]);
        imshow(d.img_specified, 'Parent', ax3);
        title(ax3, 'Hasil Specification', 'Color', lbl_fg, 'FontSize', 10);
        
        ax4 = axes('Parent', fig, 'Position', [0.03 0.06 0.30 0.35]);
        plot_rgb_hist(ax4, d.img_original, 'Histogram Asli');
        
        ax5 = axes('Parent', fig, 'Position', [0.36 0.06 0.30 0.35]);
        plot_rgb_hist(ax5, d.img_reference, 'Histogram Referensi');
        
        ax6 = axes('Parent', fig, 'Position', [0.69 0.06 0.30 0.35]);
        plot_rgb_hist(ax6, d.img_specified, 'Histogram Hasil');
    end

end
